lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,339 @@
+  // Get the execution state for the run.
+  std::shared_ptr<ExecutionState> executionState = executionStates_[runId];
+
+  // If execution has already failed due to another node, don't bother running
+  // this one.
+  if (executionState->getResultCode() != ResultCode::Executed) {
+    // Mark the node as no longer execution.
+    executionState->decrementInflightNodes();
+    return;
+  }
+
+  // Get the DeviceManager that can run the node.
+  auto it = deviceManagers_.find(node->deviceID);
+
+  if (it == deviceManagers_.end()) {
+    // Mark the node as no longer executing.
+    executionState->decrementInflightNodes();
+    executionState->setResultCode(ResultCode::Failed);
+    return;
+  }
+
+  std::shared_ptr<DeviceManager> deviceManager = it->second;
+
+  // Get the Context containing all of the inputs for the node.
+  std::unique_ptr<Context> nodeCtx =
+      executionState->getUniqueNodeContextPtr(node);
+  // Run the node using the DeviceManager.
+  deviceManager->runFunction(
+      node->name, std::move(nodeCtx),
+      [this, runId, node](RunIdentifierTy id, ResultCode resultCode,
+                          std::unique_ptr<Context> resultCtx) {
+        // Immediately move the handling of the result onto threadPool_ to
+        // avoid doing work on the DeviceManager thread.
+        std::shared_ptr<Context> sharedResultCtx = std::move(resultCtx);
+        this->threadPool_.submit(
+            [this, runId, node, resultCode, sharedResultCtx]() {
+              this->handleDeviceManagerResult(runId, resultCode,
+                                              sharedResultCtx, node);
+            });
+      });
+}
+
+void ThreadPoolExecutor::propagateOutputPlaceholders(
+    RunIdentifierTy runId, std::shared_ptr<Context> ctx) {
+  // Get the execution state for the run.
+  std::shared_ptr<ExecutionState> executionState = executionStates_[runId];
+  // Copy all of the Placeholders in ctx into the result Context for the run.
+  for (const auto &phTensorPair : ctx->pairs()) {
+    auto *placeholder = phTensorPair.first;
+    const auto *tensor = phTensorPair.second;
+
+    executionState->insertIntoResultCtx(placeholder, tensor->clone());
+  }
+}
+
+void ThreadPoolExecutor::handleDeviceManagerResult(RunIdentifierTy runId,
+                                                   ResultCode resultCode,
+                                                   std::shared_ptr<Context> ctx,
+                                                   const DAGNode *node) {
+  // Get the execution state for the run.
+  std::shared_ptr<ExecutionState> executionState = executionStates_[runId];
+
+  if (resultCode != ResultCode::Executed) {
+    // If the DeviceManager failed to execute the node, call the callback
+    // provided when ThreadPoolExecutor::run() was called, clean up the
+    // lock and execution state for the run and exit.
+    DoneCbTy cb = executionState->getCallback();
+    cb(runId, resultCode, nullptr);
```

Comment:
Pass the context back to the caller.

Reply:
There is no "complete: `Context` to pass back here. The result `Context` being prepared in `ExecutionState` will not have all of the expected `Placeholders` in it.

Reply:
Didn't they pass some context into us though? The one in `run()` - can we pass it back?

Reply:
I misunderstood - turns out that the `Context` given in `run` contains both input and output `Placeholders`. So instead of creating a new result `Context` in the `ExecutionState` class, I should move this one into it and return it whenever the callback is called.